compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1283,6 +1283,34 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return tcx.ty_error();
         }
 
+        let is_const = match &element.kind {
+            hir::ExprKind::ConstBlock(..) => true,
+            hir::ExprKind::Path(qpath) => {
+                let res = self.typeck_results.borrow().qpath_res(qpath, element.hir_id);
+                matches!(
+                    res,
+                    Res::Def(DefKind::Const | DefKind::AssocConst | DefKind::AnonConst, _)
+                )
+            }
+            _ => false,
+        };
+
+        if !is_const {
+            let is_const_fn = match element.kind {
+                hir::ExprKind::Call(func, _args) => match *self.node_ty(func.hir_id).kind() {
+                    ty::FnDef(def_id, _) => tcx.is_const_fn(def_id),
+                    _ => false,
+                },
+                _ => false,
+            };
+
+            if count.try_eval_usize(tcx, self.param_env).map_or(true, |len| len > 1) {
+                let lang_item = self.tcx.require_lang_item(LangItem::Copy, None);
+                let code = traits::ObligationCauseCode::RepeatElementCopy { is_const_fn };
+                self.require_type_meets(element_ty, element.span, code, lang_item);
```

Comment:
I don't object to this code, but I'm a bit confused what's going on here -- it seems like, in the MIR type check, we enforce that `T: Copy` unless we can statically see that the length is either 0 or 1 (which makes sense). But here, we have this `if !is_const` check that says (iiuc) "if this is not a constant value being repeated, require `T: Copy`". Why doesn't the MIR check have a similar "if not a constant value" thing going on? Or does it, and I just missed it? Perhaps it gets lowered differently in that case?